course-content/31_exploring_data_structures.qmd

```diff
@@ -0,0 +1,447 @@
+
+<center>
+    <img src="images/exploring-data-structures/join_types.png" 
+    width=700 style="margin:auto"/>
+    <p style="text-align: center">
+       Types of Joins
+    </p>
+    <br/>
+</center>
+
+The `.merge()` method (https://pandas.pydata.org/docs/reference/api/pandas.merge.html) works similarly to how table joining works in `SQL` or how the `VLOOKUP` function works in Excel. One needs to specify both dataframes, the key variables on which to join them, and the kind of join desired. 
+
+So let's look at the example below to see how it all comes together in code. 
+
+```{python}
+combined_df = df.merge(countrysize, left_on='country', right_on='nation', how='left')
+combined_df.head(2)
+```
+
+
+When you run the code above you will notice that both the `nation` key column and the `area_square_kms` column have been joined together in the new `combined_df` object. One can keep that nation column in there for control purposes, or it can be removed by using the `.drop()` method we used earlier:
+
+combined_df = combined_df.drop('nation', axis=1) 
+
+Another innovative way to put “merge” data together in pandas is with the `.concat()` function. Conceptually you can think if it like “stacking” two data objects on top of each other or side-by-side as shown in the diagram below.
+
+
+<center>
+    <img src="images/exploring-data-structures/concat.png" 
+    width=700 style="margin:auto"/>
+    <p style="text-align: center">
+       Ways to stack data
+    </p>
+    <br/>
+</center>
+
+
+
+To illustrate, let’s fetch two simple dataframes. Each contains the average scores for three subjects by year for two separate schools.
+
+```{python}
+school1_url  = "https://raw.githubusercontent.com/bcgov/ds-intro-to-python/main/data/school1.csv"
+school2_url  = "https://raw.githubusercontent.com/bcgov/ds-intro-to-python/main/data/school2.csv"
+school1_df = pd.read_csv(school1_url)
+school2_df = pd.read_csv(school2_url)
+```
+
+Let's take a quick look at the two dataframes.
+
+```{python}
+school1_df
+```
+
+```{python}
+school2_df
+```
+
+We call `.concat()` and pass in the objects that we want to stack vertically. This is a similiar operation to `union` in `SQL`.
+
+```{python}
+vertical_stack_df = pd.concat([school1_df, school2_df])
+vertical_stack_df
+```
+
+It is also possible to stack the data horizontally. Here it is necessary to specify the columnar axis (axis=1) as the default setting is for rows (axis=0).
+
+```{python}
+horizontal_stack = pd.concat([school1_df, school2_df], axis=1)
+horizontal_stack
+```
+
+This introduction only scratches the surface of how to leverage this way of joining datasets together. But it can be a powerful tool in the toolkit for the right use case. More detail found here: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.concat.html
```

Comment:
```suggestion
This introduction only scratches the surface of how to leverage this way of joining datasets together. But it can be a powerful tool in the toolkit for the right use case. [More detail found here.](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.concat.html)
```